crates/router/src/core/api_keys.rs

```diff
@@ -154,11 +162,86 @@ pub async fn create_api_key(
         &[metrics::request::add_attributes("merchant", merchant_id)],
     );
 
+    // Add process to process_tracker for email reminder, only if expiry is set to future date
+    #[cfg(feature = "email")]
+    {
+        if api_key.expires_at.is_some() {
+            let expiry_reminder_days = state.conf.api_keys.expiry_reminder_days.clone();
+
+            add_api_key_expiry_task(store, &api_key, expiry_reminder_days)
+                .await
+                .into_report()
+                .change_context(errors::ApiErrorResponse::InternalServerError)
+                .attach_printable("Failed to insert API key expiry reminder to process tracker")?;
+        }
+    }
+
     Ok(ApplicationResponse::Json(
         (api_key, plaintext_api_key).foreign_into(),
     ))
 }
 
+// Add api_key_expiry task to the process_tracker table.
+// Construct ProcessTrackerNew struct with all required fields, and schedule the first email.
+// After first email has been sent, update the schedule_time based on retry_count in execute_workflow().
+#[cfg(feature = "email")]
+#[instrument(skip_all)]
+pub async fn add_api_key_expiry_task(
+    store: &dyn StorageInterface,
+    api_key: &ApiKey,
+    expiry_reminder_days: Vec<u8>,
+) -> Result<(), errors::ProcessTrackerError> {
+    let current_time = common_utils::date_time::now();
+    let api_key_expiry_tracker = &storage::ApiKeyExpiryWorkflow {
+        key_id: api_key.key_id.clone(),
+        merchant_id: api_key.merchant_id.clone(),
+        // We need API key expiry too, because we need to decide on the schedule_time in
+        // execute_workflow() where we won't be having access to the Api key object.
+        api_key_expiry: api_key.expires_at,
+        expiry_reminder_days: expiry_reminder_days.clone(),
+    };
+    let api_key_expiry_workflow_model = serde_json::to_value(api_key_expiry_tracker)
+        .into_report()
+        .change_context(errors::ApiErrorResponse::InternalServerError)
+        .attach_printable_lazy(|| {
+            format!("unable to serialize API key expiry tracker: {api_key_expiry_tracker:?}")
+        })?;
+
+    let schedule_time = api_key.expires_at.map(|expires_at| {
+        expires_at.saturating_sub(time::Duration::days(i64::from(expiry_reminder_days[0])))
+    });
+    let process_tracker_entry = storage::ProcessTrackerNew {
+        id: generate_task_id_for_api_key_expiry_workflow(api_key.key_id.as_str()),
+        name: Some(String::from("API_KEY_EXPIRY")),
+        tag: vec![String::from("API_KEY")],
+        runner: Some(String::from("API_KEY_EXPIRY_WORKFLOW")),
```

Comment:
Can we make these values constants?

Reply:
These values are specific to a workflow (here its api_key_expiry_workflow). We have these values specific in other workflows as well (like refund, payment_sync). But those values are also kept inside their own files. I think we can keep it in their own files itself to give better context of the workflow. what do u suggest?

Reply:
Yes, I'm fine with having these values in the `api_key_expiry` workflow file, as constants.

Reply:
Cool